windows/lcow_runtime.go

```diff
@@ -0,0 +1,426 @@
+	return lcowPluginID
+}
+
+func (r *LCOWRuntime) Create(ctx context.Context, id string, opts runtime.CreateOpts) (runtime.Task, error) {
+	namespace, err := namespaces.NamespaceRequired(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := typeurl.UnmarshalAny(opts.Spec)
+	if err != nil {
+		return nil, err
+	}
+	spec := s.(*specs.Spec)
+
+	var createOpts *hcsshimopts.CreateOptions
+	if opts.Options != nil {
+		o, err := typeurl.UnmarshalAny(opts.Options)
+		if err != nil {
+			return nil, err
+		}
+		createOpts = o.(*hcsshimopts.CreateOptions)
+	} else {
+		createOpts = &hcsshimopts.CreateOptions{}
+	}
+
+	if createOpts.TerminateDuration == 0 {
+		createOpts.TerminateDuration = defaultTerminateDuration
+	}
+
+	return r.newTask(ctx, namespace, id, spec, opts.IO, createOpts)
+}
+
+func (r *LCOWRuntime) Get(ctx context.Context, id string) (runtime.Task, error) {
+	return r.tasks.Get(ctx, id)
+}
+
+func (r *LCOWRuntime) Tasks(ctx context.Context) ([]runtime.Task, error) {
+	return r.tasks.GetAll(ctx)
+}
+
+func (r *LCOWRuntime) Delete(ctx context.Context, t runtime.Task) (*runtime.Exit, error) {
+	wt, ok := t.(*task)
+	if !ok {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "not an LCOW task")
+	}
+
+	// TODO(mlaventure): stop monitor on this task
+
+	var (
+		err      error
+		state, _ = wt.State(ctx)
+	)
+	switch state.Status {
+	case runtime.StoppedStatus:
+		fallthrough
+	case runtime.CreatedStatus:
+		// if it's stopped or in created state, we need to shutdown the
+		// container before removing it
+		if err = wt.stop(ctx); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, errors.Wrap(errdefs.ErrFailedPrecondition,
+			"cannot delete a non-stopped task")
+	}
+
+	var rtExit *runtime.Exit
+	if p := wt.getProcess(t.ID()); p != nil {
+		ec, ea, err := p.ExitCode()
+		if err != nil {
+			return nil, err
+		}
+		rtExit = &runtime.Exit{
+			Pid:       wt.pid,
+			Status:    ec,
+			Timestamp: ea,
+		}
+	} else {
+		rtExit = &runtime.Exit{
+			Pid:       wt.pid,
+			Status:    255,
+			Timestamp: time.Now(),
+		}
+	}
+
+	wt.cleanup()
+	r.tasks.Delete(ctx, t)
+
+	r.publisher.Publish(ctx,
+		runtime.TaskDeleteEventTopic,
+		&eventsapi.TaskDelete{
+			ContainerID: wt.id,
+			Pid:         wt.pid,
+			ExitStatus:  rtExit.Status,
+			ExitedAt:    rtExit.Timestamp,
+		})
+
+	// We were never started, return failure
+	return rtExit, nil
+}
+
+func (r *LCOWRuntime) newTask(ctx context.Context, namespace, id string, spec *specs.Spec, io runtime.IO, createOpts *hcsshimopts.CreateOptions) (*task, error) {
+	var (
+		err  error
+		pset *pipeSet
+	)
+
+	if pset, err = newPipeSet(ctx, io); err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			pset.Close()
+		}
+	}()
+
+	var pid uint32
+	if pid, err = r.pidPool.Get(); err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			r.pidPool.Put(pid)
+		}
+	}()
+
+	///////// TODO: Once there is a snapshotter for LCOW, this can be deleted.
```

Comment:
Probably should remove this for now.

Reply:
Oops, I deleted this in my local converged branch but forgot to delete in this branch. Removed.

Reply:
There are a few more blocks like this. It would be good to clean them up. ;)

Reply:
They should all be cleaned up now.